### PR TITLE
Improved img_2_vec encoder

### PR DESCRIPTION
**Pull request by George3d6**
Method separation might look a bit strange since this is part of the prototype PR for trainable encoders. However, I think this contains a few relevant tidbits we should merge regardless:

* Removed deprecated logic and cleaned up the code + simplified it a bit
* Added support for 4 channel images (e.g. PNGs with transparency) and grayscale images (granted, by converting to 3-channel RGB, maybe a better approach exists, but better than nothing)
* Made the tests work without local images and moved them to the current test structure
* Disabled that ever-annoying PIL logger

### CODE REVIEW

File: lightwood/encoders/image/helpers/img_to_vec.py [hunk 1]
```diff
@@ -1,141 +1,75 @@
...
+class Img2Vec(nn.Module):
 
-    def __init__(self, model='resnet-18', layer='default', layer_output_size=512):
+    def __init__(self, model, layer='default'):
         """ Img2Vec
         :param cuda: If set to True, will run forward pass on GPU
         :param model: String name of requested model
         :param layer: String or Int depending on model.  See more docs: https://github.com/christiansafka/img2vec.git
         :param layer_output_size: Int depicting the output size of the requested layer
         """
```
Comment (paxcema) [inline lightwood/encoders/image/helpers/img_to_vec.py:25]: ```cuda``` and ```layer_output_size``` param info. is obsolete

File: lightwood/encoders/image/helpers/img_to_vec.py [hunk 2]
```diff
@@ -1,141 +1,75 @@
...
-                elif self.model_name in ('resnet-18', 'resnext-50'):
-                    return my_embedding.numpy()[0, :, 0, 0]
 
-    def _get_model_and_layer(self, model_name, layer):
+    def _get_model(self, layer):
         """ Internal method for getting layer from model
         :param model_name: model name such as 'resnet-18'
         :param layer: layer as a string for resnet-18 or int for alexnet
         :returns: pytorch model, selected layer
         """
```
Comment (paxcema) [inline lightwood/encoders/image/helpers/img_to_vec.py:59]: ```model_name``` in docstring obsolete

Comment (paxcema) [inline lightwood/encoders/image/helpers/img_to_vec.py:59]: ```model_name``` in docstring obsolete

File: lightwood/encoders/image/helpers/img_to_vec.py [hunk 3]
```diff
@@ -1,141 +1,75 @@
...
+            self.layer_output_size = 512
+            model = torch.nn.Sequential(*list(models.resnext50_32x4d(pretrained=True).children())[:-1] , ChannelPoolAdaptiveAvg1d(output_size=512))
+
+        elif self.model_name == 'resnext-50':
+            self.layer_output_size = 2048
+            model = torch.nn.Sequential(*list(models.resnext50_32x4d(pretrained=True).children())[:-1])
 
         else:
-            raise KeyError('Model %s was not found' % model_name)
+            raise Exception(f'Image encoding model {model_name} was not found')
```
Comment (paxcema) [inline lightwood/encoders/image/helpers/img_to_vec.py:73]: Should be ```{self.model_name}```

File: lightwood/encoders/image/helpers/img_to_vec.py [hunk 4]
```diff
@@ -1,141 +1,75 @@
...
-            model = models.mobilenet_v2(pretrained=True)
-            if layer == 'default':
-                layer = model._modules.get('classifier')
-                self.layer_output_size = 1000
-            else:
-                layer = model._modules.get(layer)
-
-            return model, layer
+        if self.model_name == 'resnet-18':
+            self.layer_output_size = 512
```
Comment (paxcema) [inline lightwood/encoders/image/helpers/img_to_vec.py:61]: Is ```self.layer_output_size``` used anywhere?

File: lightwood/encoders/image/img_2_vec.py [hunk 5]
```diff
@@ -1,79 +1,97 @@
...
+                self.model = Img2Vec(model='resnext-50')
+            else:
+                self.model = Img2Vec()
+        self._prepared = True
+
+    def prepare(self, images):
+        img_tensor_arr = []
+        for image in images:
+            if image is not None:
+                #logging.setLevel(logging.ERROR)
```
Comment (paxcema) [inline lightwood/encoders/image/img_2_vec.py:56]: Leftover?

File: lightwood/encoders/image/img_2_vec.py [hunk 6]
```diff
@@ -1,79 +1,97 @@
...
+                    img = Image.open(image)
+                #logging.setLevel(logging.DEBUG)
+
+                # In order for Normalize to work with 4-channel pngs and grayscale
+                img = img.convert('RGB')
+
+                img_tensor = self._scaler(img)
+                img_tensor = self._to_tensor(img_tensor)
+                img_tensor = self._normalize(img_tensor)
+                img_tensor = img_tensor
```
Comment (paxcema) [inline lightwood/encoders/image/img_2_vec.py:70]: This line is not needed.

File: lightwood/encoders/image/img_2_vec.py [hunk 7]
```diff
@@ -1,79 +1,97 @@
...
+            :images : list of images, each image is a path image(ToDO: url to image also need to be included)
+            :return: a torch.floatTensor
+        """
+        if not self._prepared:
+            raise Exception('You need to call "prepare_encoder" before calling "encode" or "decode".')
+
+        img_tensors = self.prepare(images)
+        vec_arr = []
+        for img_tensor in img_tensors:
+            self.model.eval()
```
Comment (paxcema) [inline lightwood/encoders/image/img_2_vec.py:90]: The ```.eval()``` and ```.no_grad()``` calls should go before the loop I think

Comment (George3d6) [reply lightwood/encoders/image/img_2_vec.py:90]: 100%, that was a terrible mistake

File: lightwood/encoders/image/helpers/img_to_vec.py (hunk 1 above)
Comment (George3d6) [reply lightwood/encoders/image/helpers/img_to_vec.py:22]: Rsolved but github is not picking it up

File: lightwood/encoders/image/img_2_vec.py [hunk 8]
```diff
@@ -1,79 +1,95 @@
...
+    def prepare(self, images):
+        img_tensor_arr = []
+        for image in images:
+            if image is not None:
+                if image.startswith('http'):
+                    response = requests.get(image)
+                    img = Image.open(BytesIO(response.content))
+                else:
+                    img = Image.open(image)
+                #logging.setLevel(logging.DEBUG)
```
Comment (paxcema) [inline lightwood/encoders/image/img_2_vec.py:61]: Leftover

**Pull request merged**